frame/assets/src/functions.rs

```diff
@@ -652,71 +658,124 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
-	///   account is the owner of that asset. Can be used for authorization checks.
-	pub(super) fn do_destroy(
+	/// Start the process of destroying an asset, by setting the asset status to Destroying, and
+	/// emitting the DestructionStarted event.
+	pub(super) fn do_start_destroy(
 		id: T::AssetId,
-		witness: DestroyWitness,
 		maybe_check_owner: Option<T::AccountId>,
-	) -> Result<DestroyWitness, DispatchError> {
-		let mut dead_accounts: Vec<T::AccountId> = vec![];
-
-		let result_witness: DestroyWitness = Asset::<T, I>::try_mutate_exists(
-			id,
-			|maybe_details| -> Result<DestroyWitness, DispatchError> {
-				let mut details = maybe_details.take().ok_or(Error::<T, I>::Unknown)?;
+	) -> DispatchResult {
+		let _ =
+			Asset::<T, I>::try_mutate_exists(id, |maybe_details| -> Result<(), DispatchError> {
+				let mut details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
 				if let Some(check_owner) = maybe_check_owner {
 					ensure!(details.owner == check_owner, Error::<T, I>::NoPermission);
 				}
-				ensure!(details.accounts <= witness.accounts, Error::<T, I>::BadWitness);
-				ensure!(details.sufficients <= witness.sufficients, Error::<T, I>::BadWitness);
-				ensure!(details.approvals <= witness.approvals, Error::<T, I>::BadWitness);
+				ensure!(details.is_frozen, Error::<T, I>::NotFrozen);
+				details.status = AssetStatus::Destroying;
+
+				Self::deposit_event(Event::DestructionStarted { asset_id: id });
+				Ok(())
+			})?;
+		Ok(())
+	}
+
+	/// Destroy accounts associated with a given asset up to the max (T::RemoveKeysLimit).
+	///
+	/// Each call emits the `Event::DestroyedAccountss` event.
+	pub(super) fn do_destroy_accounts(id: T::AssetId) -> Result<u32, DispatchError> {
+		let mut dead_accounts: Vec<T::AccountId> = vec![];
+		let mut removed_accounts = 0;
+		let _ =
+			Asset::<T, I>::try_mutate_exists(id, |maybe_details| -> Result<(), DispatchError> {
+				let mut details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
+				ensure!(details.is_frozen, Error::<T, I>::NotFrozen);
+				// Should only destroy accounts while the asset is being destroyed
+				ensure!(details.status == AssetStatus::Destroying, Error::<T, I>::LiveAsset);
 
 				for (who, v) in Account::<T, I>::drain_prefix(id) {
-					// We have to force this as it's destroying the entire asset class.
-					// This could mean that some accounts now have irreversibly reserved
-					// funds.
 					let _ = Self::dead_account(&who, &mut details, &v.reason, true);
 					dead_accounts.push(who);
+					removed_accounts = removed_accounts.saturating_add(1);
+					if removed_accounts >= T::RemoveKeysLimit::get() {
+						break
+					}
 				}
-				debug_assert_eq!(details.accounts, 0);
-				debug_assert_eq!(details.sufficients, 0);
+
+				Self::deposit_event(Event::DestroyedAccounts {
+					asset_id: id,
+					accounts_destroyed: removed_accounts as u32,
+					accounts_remaining: details.accounts as u32,
+				});
```

Comment:
still feels strange to me why its not at the end, after `T::Freezer::died(id, &who);`

Reply:
But you're right, it's likely safer there as well. I just won't have access to details.account from outside. But i'll just copy it into another variable.